File: 07-system/cpu/sched/cfs_demo.cpp
// cfs_demo.cpp
#include <cerrno>
#include <csignal>
#include <cstring>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <sys/resource.h>
#include <sys/sysinfo.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sched.h>
#include <unistd.h>
#include <vector>

struct Child {
    std::string name;
    int nice_value;
    pid_t pid;
};


void bind_to_cpu0() {
    cpu_set_t set;
    CPU_ZERO(&set);
    CPU_SET(0, &set);

    if (sched_setaffinity(0, sizeof(set), &set) != 0) {
        std::cerr << "sched_setaffinity failed: " << strerror(errno) << "\n";
        _exit(1);
    }
}

void busy_loop(int nice_value) {
    if (setpriority(PRIO_PROCESS, 0, nice_value) != 0) {
        std::cerr << "setpriority failed: " << strerror(errno) << "\n";
        _exit(1);
    }

    bind_to_cpu0();

    volatile unsigned long long x = 0;
    while (true) {
        ++x;
    }
}


double read_cpu_time(pid_t pid) {
    std::ifstream file("/proc/" + std::to_string(pid) + "/stat");
    if (!file) {
        throw std::runtime_error("failed to open /proc/<pid>/stat");
    }

    std::string line;
    std::getline(file, line);

    std::istringstream iss(line);
    std::string field;

    long utime = 0;
    long stime = 0;

    for (int i = 1; iss >> field; ++i) {
        if (i == 14) {
            utime = std::stol(field);
        } else if (i == 15) {
            stime = std::stol(field);
            break;
        }
    }

    long ticks = sysconf(_SC_CLK_TCK);
    return static_cast<double>(utime + stime) / ticks;
}

int main() {
    std::vector<Child> children;

    std::vector<std::pair<std::string, int>> configs = {
        {"normal", 0},
        {"low-priority", 10},
    };

    for (const auto& [name, nice_value] : configs) {
        pid_t pid = fork();

        if (pid < 0) {
            std::cerr << "fork failed: " << strerror(errno) << "\n";
            return 1;
        }

        if (pid == 0) {
            busy_loop(nice_value);
        }

        children.push_back({name, nice_value, pid});
    }

    sleep(1);

    std::vector<double> start;
    for (const auto& child : children) {
        start.push_back(read_cpu_time(child.pid));
    }

    int duration = 10;
    std::cout << "run " << duration << "s...\n\n";
    sleep(duration);

    std::vector<double> end;
    for (const auto& child : children) {
        end.push_back(read_cpu_time(child.pid));
    }

    for (size_t i = 0; i < children.size(); ++i) {
        double cpu_time = end[i] - start[i];

        std::cout << children[i].name
                  << " nice=" << children[i].nice_value
                  << " pid=" << children[i].pid
                  << " cpu_time=" << cpu_time << "s\n";
    }

    for (const auto& child : children) {
        kill(child.pid, SIGKILL);
        waitpid(child.pid, nullptr, 0);
    }

    return 0;
}

File: 07-system/cpu/sched/cfs_demo.py
# cfs_demo.py
import os
import time
import signal
import resource

def busy_loop(name, nice_value):
    os.nice(nice_value)

    # 绑到同一个 CPU，避免它们跑到不同核上
    os.sched_setaffinity(0, {0})

    x = 0
    while True:
        x += 1

def read_cpu_time(pid):
    # /proc/<pid>/stat 里第 14、15 列是 utime/stime
    with open(f"/proc/{pid}/stat") as f:
        fields = f.read().split()

    utime = int(fields[13])
    stime = int(fields[14])
    ticks = os.sysconf(os.sysconf_names["SC_CLK_TCK"])

    return (utime + stime) / ticks

children = []

for name, nice_value in [("normal", 0), ("low-priority", 10)]:
    pid = os.fork()
    if pid == 0:
        busy_loop(name, nice_value)
    else:
        children.append((name, nice_value, pid))

time.sleep(1)

start = {
    pid: read_cpu_time(pid)
    for _, _, pid in children
}

duration = 10
print(f"run {duration}s...\n")
time.sleep(duration)

end = {
    pid: read_cpu_time(pid)
    for _, _, pid in children
}

for name, nice_value, pid in children:
    cpu_time = end[pid] - start[pid]
    print(f"{name:12s} nice={nice_value:2d} pid={pid} cpu_time={cpu_time:.2f}s")

for _, _, pid in children:
    os.kill(pid, signal.SIGKILL)
